### ISSUE TITLE
📎 Interactive yes/no when suggesting CLI commands in a TTY

### ISSUE THREAD

**Issue by sebmck** [romefrontend/rome#974]
Inspired by https://twitter.com/goshakkk/status/1291161477353799681

We can ask the user if they really meant the command that was suggest in the console when in a TTY.

We can perform some validation before on the suggested command flags too to make sure that accepting it wont result in more errors.

**Comment by adityapurwa** [comment]
Hi, I would like to try and help to implement this feature. Are there any constraints that you had in mind? I imagined a scenario of:

In case when there's only one possible command
```
$ rome chekc
> did you mean `rome check`? [Y/n]:
```
When the user pressed enter, we run the command, when they pressed a possible cancel key (backspace? esc? q?) - we immediately return the control back to the user.

When there are multiple commands:

```
$ rome losg
> did you mean:
1 ) `rome logs` <-
2 ) `rome lsp`
   [ESC] To cancel
```

I think we also should consider the granularity, e.g:

```
$ rome congif enble
> did you mean:
1 ) `rome config enable`
2 ) `rome config disable`
```

We would have a maps of words that are possible at the root level, then on each subcommand.

```
{
 ...
 'config' : {
    ...
    'enable': {},
    'disable': {}
    ...
  }
 ...
}
```

Let me know what you think about it and if I can be of any help. Thanks!

**Comment by sebmck** [comment]
Yeah that sounds right. Feel free to take it on! The changes would just be confined to `internal/cli-flags/Parser.ts`. I'm not really sure how it would all work. We verify that a command was specified (it could just be only arguments) in `commandRequired`.

You actually wouldn't need any existing map to produce the suggested commands, in fact, we already have a `findClosestStringMatch` call in `commandRequired` that uses `this.commands.keys()` which contains:

```
[
  'start',                'develop',
  'stop',                 'run',
  'restart',              'status',
  'lsp',                  '_evict',
  '_moduleSignature',     '_projectDump',
  'analyzeDependencies',  'bundle',
  'cache dir',            'cache clear',
  'check',                'ci',
  'compile',              'config location',
  'config disable',       'config enable',
  'config push',          'config set',
  'config set-directory', 'format',
  'init',                 'json',
  'noop',                 'parse',
  'publish',              'resolve',
  'test',                 'recover apply',
  'recover clear',        'recover diff',
  'recover dir',          'recover list',
  'recover pop',          'rage',
  'logs'
]
```

The only reason it doesn't suggest anything for `rome congif enble` is because we don't pass in a min rating for the string comparison and it defaults to `0.8` which is less than the closest match.

We already have all the pieces necessary for the actual interactivity part. On `Parser` instance there is a `reporter` property that contains an instance of [`Reporter`](https://github.com/romefrontend/rome/blob/8bffca3bf752865c6ad02f66887cdd76b9f63677/internal/cli-reporter/Reporter.ts) which is what we use for any console printing. The `select` and `confirm` methods should be enough to cover the two scenarios you listed.

You would likely need to modify `commandRequired` to be async, and when a suggested command is chosen, clear the internal data structures (everything defined in `constructor`), call `consumeRawArgs` with the raw suggested arguments, and then rerun `init`.

`serializeCLIFlags` also only returns a complete string, so you would need to somehow get an array of arguments. The actual source text building could be pulled out from that method, although I'm not sure how tricky it would be...

Sorry if I'm throwing a lot of terms and methods/files at you, happy to clarify if you have any questions and provide more detailed instructions.

**Comment by ematipico** [comment]
Bumping this issue. There are already lot of information in order to proceed with the implementation. It might not be a good first issue, but the changes are confined to just one file.

**Comment by noahsb** [comment]
If this is still available in a couple of days I'd love to work on it!

**Comment by UsamaHameed** [comment]
I started working on this and made plenty of progress. Just need to figure out how to cleanup and init the cli. 💯
